### ISSUE TITLE
Localize Text

### ISSUE THREAD

**Issue by titis14** [UnchartedBull/OctoDash#914]
Hello I installed octodash on my pi with a 7 inch screen in 1024x800 I am delighted with this application. I would like to know if the French translation is possible? I would like to contribute to translate the application. Is this easily done from a language file or is everything hard in the code?
Thank you

**Comment by UnchartedBull** [comment]
Currently everything is hardcoded unfortunately. But having an i18n translation file is definitely on the list of improvements now. I'm changing the issues title and would come back to you for a french translation once this is done. Would this work for you?

**Comment by titis14** [comment]
Hello,
Ok great I'm waiting for your return to be able to do the translation :)

**Comment by Der-WU** [comment]
Hi there.
I can translate it to German if it rdy. :)

**Comment by Protoncek** [comment]
I can translate to Slovenian...

**Comment by Zero-Grav** [comment]
Hi 
Could you please give me a language translation tutorial？！

**Comment by TroisSix** [comment]
Also interested to help traduct in french :)

**Comment by pciavald** [comment]
I have located and translated all hard-coded strings of the app to french, i can implement this feature if the correct structure is implemented, or at least advised (long time since i worked on an angular app, i would not be the right person to choose the plugins and file architecture).

**Comment by pciavald** [comment]
If no indication is given, i will start tomorrow. We have just rebased our branch to your latest main, and most changes are in conflict so i'll be faster implementing it correctly than fixing the merge :)

**Comment by UnchartedBull** [comment]
Didn't had a thorough look at the possibilities yet, but `ngx-translate` seems promising, but I'm open to other libraries / the default i18n localisation as well. It should be able to:

- Handle untranslated content (i.e. adding a new feature with no translation available yet)
- Should work in serve mode (at least in english)
- Don't dramatically increase build times

If you find something let me know and I can have a look at that :)

**Comment by pciavald** [comment]
I'm following this at the moment: https://angular.io/guide/i18n

**Comment by pciavald** [comment]
Mostly done, will PR tomorrow

**Comment by UnchartedBull** [comment]
Thanks to @pciavald this has been implemented and a french translation is also available. If anyone of you want to add another translation please feel free to do so. A how-to is available in the CONTRIBUTING file. Thanks :)

**Comment by Max3155** [comment]
> Hi there.
> I can translate it to German if it rdy. :)

Hi, can you help me with the translation to germany? I need this for a projekt in a few days.

**Comment by Max3155** [comment]
Or can someone else help me who is familiar with it?

**Comment by jneilliii** [comment]
Are you having issues with these instructions? https://github.com/UnchartedBull/OctoDash/blob/main/CONTRIBUTING.md#add-a-new-language

**Comment by Max3155** [comment]
Yes I can't follow the instructions properly somehow. Is there a video or something?

**Comment by jneilliii** [comment]
none that I'm aware of. Have you at least got to the point where you have a german xlf file?

**Comment by Max3155** [comment]
No, unfortunately not. The npm somehow doesn't work either. No matter what command I enter nothing comes back

**Comment by jneilliii** [comment]
you can just copy this file and fill in the "target" sections for the german equivalent. https://raw.githubusercontent.com/UnchartedBull/OctoDash/main/src/locale/messages.fr.xlf

**Comment by jneilliii** [comment]
once you have that I might be able to build you a version tonight that will work.

**Comment by Max3155** [comment]
Okay I understand that. How do I go on then?

**Comment by Max3155** [comment]
okay thank you very much

**Comment by Max3155** [comment]
Can you then show me how it works?

**Comment by jneilliii** [comment]
I've never done it, so would have to figure that part out, but have a local development setup of OctoDash at home. Pretty sure it's just a bunch of copying/pasting of source files in OctoDash to add the language.

**Comment by Max3155** [comment]
okay, I'll translate it now

**Comment by jneilliii** [comment]
Just a heads-up, I won't be home for about 6 hours, but can definitely try the process for you then. I'll keep screenshots and such to document the process for others.

**Comment by Max3155** [comment]
Okay, Thank you

**Comment by Max3155** [comment]
Okay, I'm ready with the file.

**Comment by UnchartedBull** [comment]
Please share the file here (just drag&drop) or create a PR

**Comment by Max3155** [comment]
Okay it's not perfect yet, but i will optimize it when it works


[locale_DE.zip](https://github.com/UnchartedBull/OctoDash/files/6918520/locale_DE.zip)

**Comment by Max3155** [comment]
Could you send me a short video recording in which you translate it with the file?

**Comment by jneilliii** [comment]
I think the line needs to be updated.

```
<file original="ng2.template" datatype="plaintext" source-language="en" target-language="fr">
```

to 

```
<file original="ng2.template" datatype="plaintext" source-language="en" target-language="de">
```

**Comment by UnchartedBull** [comment]
https://github.com/UnchartedBull/OctoDash/tree/feature/german-translation I created this branch. Everything is working there. You just need to execute `ng serve --configuration=de & npm run electron:serve:big` to get the development server up and running. I'll attach a deb file once the build is through.

On a site note: The translation do need a bit of work, not a huge fan of "Sie" and I think "Du" might be better suited for something like OctoDash, plus some texts are fairly long, so they might not fit in their designated spots.

<img width="758" alt="Screenshot 2021-08-02 at 19 38 46" src="https://user-images.githubusercontent.com/19818061/127902075-dbc7c6a8-c205-42e4-a8b6-f7eaae3d7d2c.png">

**Comment by UnchartedBull** [comment]
@jneilliii updated that in the linked branch already.

**Comment by UnchartedBull** [comment]
https://drive.google.com/file/d/1ZEMsNEkgBzbXCeqP38lXdsOoH-Gs44Ys/view?usp=sharing here is the deb with the included german translation

**Comment by Max3155** [comment]
Okay, thank you very much, I'll test it out and then improve the translations again.

**Comment by Max3155** [comment]
what do I have to do now?

![Unbenannt](https://user-images.githubusercontent.com/88329519/127904097-1e412be3-8b8f-4d33-b494-4d26aa983224.PNG)

**Comment by Max3155** [comment]
How can I actually create such a .deb? and then how do i use it?

**Comment by jneilliii** [comment]
I have github actions on my fork of OctoDash that will build the deb files. It's part of the packaging process. Once you have the deb file, you transfer it to your pi and use the command `sudo dpkg -i ./octodash.deb` from the folder you saved to via SSH or the terminal.

**Comment by jneilliii** [comment]
`npm run pack` will package into a deb file. not sure if you have to pass additional parameters for a different language or not to that command.

**Comment by Max3155** [comment]
Okay and how i can start octodash now?

**Comment by Max3155** [comment]
![Unbenannt](https://user-images.githubusercontent.com/88329519/127912833-95e7b8ef-d193-40ca-a97d-e0ebda50dede.PNG)

**Comment by jneilliii** [comment]
did you run through the standard install script first? I would do that and then install over the version it installs with the de one. then run `sudo service getty@tty1 restart` to have it restart.

**Comment by Max3155** [comment]
Yes now it worked. Many Thanks. Can I actually personalize the symbols?

**Comment by Max3155** [comment]
Can one of you show me how I can create the .deb?

**Comment by jneilliii** [comment]
Like I mentioned before, `npm run pack` from the folder where the code is will compile the code into a deb file. I do not know if you have to pass any parameters to that for specific language or not.

**Comment by jneilliii** [comment]
https://github.com/UnchartedBull/OctoDash/blob/main/CONTRIBUTING.md#setup

**Comment by Max3155** [comment]
npm run pack always creates a .tgz file for me

**Comment by jneilliii** [comment]
that's a compressed file. should be able to be extracted. on windows I use 7 zip.

**Comment by Max3155** [comment]
Yes I know, but I need a .deb file

**Comment by Max3155** [comment]
> https://drive.google.com/file/d/1ZEMsNEkgBzbXCeqP38lXdsOoH-Gs44Ys/view?usp=sharing here is the deb with the included german translation

I will create this deb file

**Comment by jneilliii** [comment]
once you extract the tgz is there no deb file inside?

**Comment by Max3155** [comment]
yes unfortunately no deb file
![Unbenannt](https://user-images.githubusercontent.com/88329519/128058090-b0b9c4fc-195e-414c-b208-4f8127631554.PNG)

**Comment by jneilliii** [comment]
what's in the build folder?

**Comment by Max3155** [comment]
icon.png

**Comment by Max3155** [comment]
Do you know the control file for the deb creator?

**Comment by UnchartedBull** [comment]
The `package.json` file controls the deb creation. It should create a deb archive when running `npm run pack` though. Did the command return any errors? It's also not advised to run the deb file creation on the Pi itself, because it's not powerful enough and can run out of memory during the build proces.

Also please either open a new issue or continue this on the discord. We're pinging other people here for a completely unrelated issue. Thanks.

**Comment by UnchartedBull** [comment]
Discord dev or support channel probably is your best bet.
